(DOC-3716) Document Timestamp and Timespan types (#884)
puppetlabs/puppet-docs: source/puppet/5.5/lang_data_abstract.md

* `Pattern[/\A[a-z].*/]` --- matches any string that begins with a lowercase letter.
* `Pattern[/\A[a-z].*/, /\Anone\Z/]` --- matches the above **or** the exact string `"none"`.

### `Timespan`

> The `Timespan` data type matches a Timespan; [strings][], or [numbers][] that represent a time and date; or `default`, which represents an infinite point in the past or future.

Reviewer: ```suggestion
The `Timespan` data type matches Timespan values within a given range; specified with [strings][] or [numbers][] that represent duration; or `default`, which represents a positive or infinite duration.
```

Reviewer: typo, corrected - should have read "an infinite negative or positive duration" - in practical terms "+ or - the largest possible (i.e. infinite) duration of time", as in "10 minutes or longer" (the upper end is positive infinity), or "you can subtract as much time as you like" (the lower end is negative infinity).

It takes up to two parameters, and defaults to an infinite range to the past and future.

#### Parameters

The full signature for `Timespan` is:

    Timespan[ (<TIMESPAN START OR LENGTH>, (<TIMESPAN END>)) ]

Position | Parameter | Data Type | Default Value | Description
---------|-----------|-----------|---------------|------------
1 | Timespan Start or Length | String, Float, Integer, or `default` | `default` (-Infinity in a span) | Length of the timespan if passed alone, or the `from` value in a range if passed with a second parameter
2 | Timespan End | String, Float, Integer, or `default` | `default` (+Infinity), or none if only one value is passed | The `to` value in a range

`Timespan` values are interpreted depending on their format.

* A String in the format of `D-HH:MM:SS` represents a span of days (`D`), hours (`HH`), minutes (`MM`), and seconds (`SS`)
* An Integer or Float represents a number of seconds

A `Timespan` range with either end defined as `default` (infinity) is an _open range_, while any other span is a _closed range_. The span's range is inclusive.

The `Timespan` type is not enumerable.

> 

Reviewer: May want to link to the documentation for Timspan.new https://puppet.com/docs/puppet/6.0/function.html#conversion-to-timespan

#### Examples:

> * `Timespan[2]` --- a timespan of 2 seconds

Reviewer: ```suggestion
* `Timespan[2]` --- matches a Timespan value of 2 seconds
```

> * `Timespan[77.3]` --- a timespan of 1 minute, 17 seconds, and 300 milliseconds (77.3 seconds)

Reviewer: ```suggestion
* `Timespan[77.3]` --- matches a Timespan value of 1 minute, 17 seconds, and 300 milliseconds (77.3 seconds)
```

> * `Timespan['1-00:00:00', '2-00:00:00']` --- a timespan between 1 and 2 days

Reviewer: ```suggestion
* `Timespan['1-00:00:00', '2-00:00:00']` --- matches  Timespan values between 1 and 2 days
```

### `Timestamp`

> The `Timestamp` data type matches a Timestamp; [strings][] or [numbers][] that represent a time and date; or `default`, which represents an infinite point in the past or future.

Reviewer: This has the same issues as for `Timespan` - a `Timestamp` only matches `Timestamp` values.

> The `Timestamp` data type matches a Timestamp; [strings][] or [numbers][] that represent a time and date; or `default`, which represents an infinite point in the past or future.

Reviewer: ```suggestion
The `Timestamp` data type matches Timestamp values within a given range; specified with [strings][] or [numbers][] that represents the points in time limiting the range; or `default`, which represents an infinite point in the past or future
```

Reviewer: A range of time stamps are bound to points in time; from a specific date/time to another specific date/time. A range of Timestamps is a range of durations. For example - a movie may be 2h30min long (a Timespan), but that does not say on which date and time it is showing (a Timstamp), so a range of timespans may be "movies longer than 60 minutes but shorter than 4h". 

We use `default` in the type system to describe the absolute min/max values of things - we have that already in Integer and Float ranges. It works the same here; the default upper/lower range values can also be explicitly given as literal `default`. Does that make it clearer?

Reviewer: BTW: the term "time" is best avoided, since it is after all relative ;-) and it does not say if you mean a "point in time" (date/time) or a "duration of time" (length). So `Timestamp` is "When", and `Timespan` is "How Long"

> It takes up to two parameters, and defaults to an infinite range to the past and future. A `Timestamp` with one parameter represents a single point in time, while two parameters represent a range of time, with the first parameter being the `from` value and the second being the `to` value.

Reviewer: ```suggestion
It takes up to two parameters, and defaults to an infinite range to the past and future. A `Timestamp` with one parameter represents a single point in time, while two parameters represent a range of timestamps, with the first parameter being the `from` value and the second being the `to` value.
```

Reviewer: I think I answered that in one of the other questions.

2 | Range Limit | String, Float, Integer, or `default` | `default` (+Infinity), or none if only one value is passed | The `to` value in a range

`Timestamp` values are interpreted depending on their format.

* A String in the format of `YYYY-MM-DD` represents the time 00:00:00.0 UTC on the specified year (`YYYY`), month (`MM`), and day (`DD`)
* A String in the format of `YYYY-MM-DD`T`HH:MM:SS.SSS` represents a specific hour (`HH`), minute (`MM`), and decimal second (`SS.SSS`) of UTC on the specified date
* An Integer or Float represents the seconds passed since the Unix epoch
* `default` is used in ranges and represents infinity as the maximum floating point value, either negative (as the `from` value of a range) or positive (as the `to` value)

A `Timestamp` range with either end defined as `default` (infinity) is an _open range_, while a range with both ends defined as an exact timestamp is a _closed range_. Range values can be negative, and the range is inclusive.

The `Timestamp` type is not enumerable.

> 

Reviewer: May want to link to the documentation for Timspan.new https://puppet.com/docs/puppet/6.0/function.html#conversion-to-timestamp

#### Examples:

> * `Timestamp['2000-01-01T00:00:00.000', default]` --- an open range of time from the start of the 21st century to an infinte point in the future

Reviewer: ```suggestion
* `Timestamp['2000-01-01T00:00:00.000', default]` --- an open range of timestamps from the start of the 21st century to an infinte point in the future
```

> * `Timestamp['2012-10-10']` --- the exact Timestamp 2012-10-10T00:00:00.0 UTC

Reviewer: ```suggestion
* `Timestamp['2012-10-10']` --- matches the exact Timestamp 2012-10-10T00:00:00.0 UTC
```

> * `Timestamp[default, 1433116800]` --- an open range of time from an infinite point in the past to 2015-06-01T00:00:00 UTC, here expressed as seconds since the Unix epoch

Reviewer: ```suggestion
* `Timestamp[default, 1433116800]` --- an open range of timestamps from an infinite point in the past to 2015-06-01T00:00:00 UTC, here expressed as seconds since the Unix epoch
```

> * `Timestamp['2010-01-01', '2015-12-31T23:59:59.999999999']` --- a closed range of time between the start of 2010 and the end of 2015

Reviewer: ```suggestion
* `Timestamp['2010-01-01', '2015-12-31T23:59:59.999999999']` --- a closed range of timestamps between the start of 2010 and the end of 2015
```
